Improve shared key derivation and add more test vectors (#24)
chris-wood/draft-bar-cfrg-spake2plus: draft-bar-cfrg-spake2plus.md

## Key Schedule and Key Confirmation {#keys}

The protocol transcript TT, as defined in {{online-authentication}}, is unique and secret to
the participants. Both parties use TT to derive the shared symmetric secret K_main from the
protocol. The length of K_main is equal to the length of the digest output, e.g., 256 bits
for Hash() = SHA-256. The confirmation keys K_confirmP and K_confirmV, as well as the shared

> encryption key K_shared are derived from K_main.

Author: Ah, yes. Good spot!

## Key Schedule and Key Confirmation {#keys}

The protocol transcript TT, as defined in {{online-authentication}}, is unique and secret to
the participants. Both parties use TT to derive the shared symmetric secret K_main from the
protocol. The length of K_main is equal to the length of the digest output, e.g., 256 bits
for Hash() = SHA-256. The confirmation keys K_confirmP and K_confirmV, as well as the shared

> encryption key K_shared are derived from K_main. The length of K_shared can be freely chosen
> by the application.

Reviewer: ```suggestion
encryption key K_shared are derived from K_main.
```

> encryption key K_shared are derived from K_main. The length of K_shared can be freely chosen
> by the application.

Reviewer: ```suggestion
encryption key K_shared are derived from K_main. The length of K_shared is equal to the
length of the KDF output, which is defined by the ciphersuite.
```

~~~
K_main = Hash(TT)
K_confirmP || K_confirmV = KDF(nil, K_main, "ConfirmationKeys")
K_shared = KDF(nil, K_main, "SharedKey")
~~~

Neither K_main nor its derived confirmation keys are used for anything except key
derivation and confirmation and MUST be discarded after the protocol execution.

> Besides K_shared, applications MAY derive additional keys from K_main as needed.

Reviewer: ```suggestion
Applications MAY derive additional keys from K_shared as needed.
```

The length of each confirmation key is dependent on the MAC function of the chosen

> ciphersuite. For HKDF, the RECOMMENDED key length is equal to the output length of

Reviewer: ```suggestion
ciphersuite. For HMAC, the RECOMMENDED key length is equal to the output length of
```
